File: _includes/codes/talks/pattern-matching-approach/v1.rb
class ErgZone::WorkoutMode do
  def self.run(intervals)
    new(intervals).run
  end

  def initialize(intervals)
    @intervals = intervals
    @first_type = intervals[0].type
  end

  attr_reader :intervals, :first_type

  def run
    return "VariableInterval" unless same_type?

    case first_type
      when "dist" then "FixedDistSplits"
      when "time" then "FixedTimeSplits"
    end
  end

  def same_type?
    intervals.all? do |interval|
      interval.type == first_type
    end
  end
end

File: _includes/codes/talks/pattern-matching-approach/v1.ex
defmodule ErgZone.WorkoutMode do
  def run([%{type: "dist"}]) do
    "FixedDistSplits"
  end

  def run([%{type: "time"}]) do
    "FixedTimeSplits"
  end

  def run([
        %{type: t}
        | [%{type: t} | _] = tail
      ]) do
    run(tail)
  end

  def run(_intervals) do
    "VariableInterval"
  end
end
